README.md
Licensr
=======

License your projects with a nice CLI tool!

[![Dependencies Status][gemnasium-image]][gemnasium-url] [![Build Status][travis-image]][travis-url] [![Code quality][codeclimate-image]][codeclimate-url] [![Release][npm-image]][npm-url]

[![ruche logo](https://raw.githubusercontent.com/quentinrossetti/licensr/master//assets/animated.gif)](https://github.com/quentinrossetti/licensr)

Usage
---

Just run `licensr` an follow the guide!

If you choose to use *Licensr* as a node module you can use super greats
*Promises* or you can stick with the node's callbacks, it is you choice.

```js
var licensr = require('licensr');

// Super cool Promise-style
licensr.command('git config --get user.name')
  .then(function (username) {
    // on fulfill
    console.log(username);
  }, function (err) {
    // on reject
    console.error(err);
  });

// Callback-style
licensr.command('git config --get user.name', function (err, username) {
  if (err) { return console.error(err); }
  console.log(username);
});

```

Installation
---

Configure your git environement and install *Licensr* as global node module.

> *Git* is not a requirement but it will autocomplete fields for you!

```bash
git config --global user.name "John Doe"
git config --global user.email johndoe@example.com
npm install -g licensr
```

API
---

### Intro `lincensr.intro`
Displays the introduction header.

__Return values__
* `intro` The introduction text.

### Command
Run a command in the shell and get the result.

__Arguments__
* `cmd` The command to run.

__Return values__
* `stdout` The result of the command.
* `err` The error as issued by `child_process.exec`.

### Prompt License `lincensr.promptLicense`
Ask user about the license they want.

__Return values__
* `licence` A license's name.

### Prompt Info `lincensr.promptInfo`
Ask user about information required by the license.

__Arguments__
* `license` A license's name.
* `infoDefault` Default values issued by a previous command.

__Return values__
* `answers` Information about the user.

### Write
Write the license.

__Arguments__
* `data` The summary fo needed data.

__Return values__
* `err` If there is any error.
* `content` Your pimped license!

***
With :heart: from [quentinrossetti](https://github.com/quentinrossetti)

[gemnasium-url]: https://gemnasium.com/quentinrossetti/licensr
[gemnasium-image]: http://img.shields.io/gemnasium/quentinrossetti/licensr.svg
[travis-url]: https://travis-ci.org/quentinrossetti/licensr
[travis-image]: http://img.shields.io/travis/quentinrossetti/licensr.svg
[codeclimate-url]: https://codeclimate.com/github/quentinrossetti/licensr
[codeclimate-image]: http://img.shields.io/codeclimate/github/quentinrossetti/licensr.svg
[coveralls-url]: https://coveralls.io/r/quentinrossetti/licensr
[coveralls-image]: http://img.shields.io/coveralls/quentinrossetti/licensr.svg
[npm-url]: https://www.npmjs.org/package/licensr
[npm-image]: http://img.shields.io/npm/v/licensr.svg


CHANGELOG.md
Changelog
=========

### `v0.1.0` 2014-08-14

 * API: remove `prompt` in favor of `promptLicense` and `promptInfo`
 * Dependencies: use *Inquirer.js* instead of home-made functions.
 * Licenses: Add Apache-2.0, BSD-2-Clause, BSD-3-Clause, GPL-2.0, GPL-3.0,
   CDDL-1.0, EPL-1.0, LGPL-2.1, LGPL-3.0 and MPL-2.0.

### `v0.0.1` 2014-08-13

 * Initial release.

> Dates are in *ISO 8601* format (year-month-day)


test/unit.js
/* global describe, it */
'use strict';
var expect = require('chai').expect;
var licensr = require('../index');
var fixture = {
  mute: function () {},
  unmute: process.stdout.write
};

describe('API unit tests: intro', function() {

  it('should get a friendly message', function (done) {
    licensr.intro().then(function (message) {
      expect(message).to.be.a('string');
      expect(message.substr(0, 7)).to.eql('Welcome');
      done();
    });
  });

});

describe('API init test: promptLicense', function() {

  it('should responds on user input', function (done) {
    process.stdout.write = fixture.mute;
    process.nextTick(function() {
      process.stdin.emit('data', '\n');
    });
    licensr.promptLicense().then(function (answer) {
      process.stdout.write = fixture.unmute;
      expect(answer.license).to.eql('ISC');
      done();
    });
  });

});

describe('API unit tests: command', function() {

  it('should get an error on invalid command', function (done) {
    licensr.command('¤').then(null, function (err) {
      expect(err.message).to.contain('Command failed');
      done();
    });
  });

  it('should get stdout as value', function (done) {
    var cmd = (process.platform === 'win32') ? 'dir' : 'ls';
    licensr.command(cmd).then(function (output) {
      expect(output).to.be.a('string');
      done();
    });
  });

});

describe('API unit tests: write', function() {

  it('should get an error on invalid license name', function (done) {
    licensr.write({ license: '???' }).then(null, function (err) {
      expect(err.message).to.contain('Wrong license name');
      done();
    });
  });

  it('should get an error on invalid path', function (done) {
    var data = {
      license: 'ISC',
      file   : (process.platform === 'win32') ? 'D:/lic' : '/var/www/LICENSE'
    };
    licensr.write(data).then(null, function (err) {
      expect(err.message).to.be.a('string');
      done();
    });
  });

  it('should get an error on invalid filename', function (done) {
    var data = {
      license: 'ISC',
      file   : 'node_modules'
    };
    licensr.write(data).then(null, function (err) {
      expect(err.code).to.eql('EISDIR');
      done();
    });
  });

  it('should works with valid data', function (done) {
    var data = {
      name   : 'John Doe',
      email  : 'johndoe@example.com',
      years  : '2009-2014',
      license: 'Apache-2.0',
      file   : '.tmp/test/LICENSE'
    };
    licensr.write(data).then(function (license) {
      expect(license).to.be.a('string');
      done();
    });
  });

});


bin.js
#!/usr/bin/env node
'use strict';
require('colors');
var api  = require('./index');
var path = require('path');
var _    = require('underscore');

var data = {};
var infoDefault = {};

api.intro()

  .then(function (intro) {
    console.log(intro);

    // Get defaults
    api.command('git config --get user.name')
    .then(function (name) {
      infoDefault.name = name;
      return api.command('git config --get user.email');
    })
    .then(function (email) {
      infoDefault.email = email;
    })
    .catch(function (err) {
      infoDefault.name = null;
      infoDefault.email = null;
      console.error('Warning: '.yellow + 'You should install git!');
    });

    return api.promptLicense();
  })

  .then(function (answer) {
    data.license = answer.license;
    return api.promptInfo(answer.license, infoDefault);
  })

  .then(function (info) {
    data.name        = info.name;
    data.email       = info.email;
    data.software    = info.software;
    data.description = info.description;
    data.years       = info.years;
    data.file        = info.file;
    return api.write(data);
  })

  .then(function (license) {
    if (license.split('\n').length < 35) {
      console.log('\n--------\n'.grey);
      console.log(license);
      console.log('--------\n'.grey);
    }

    var warns = [ 'LGPL-2.1' ];
    if (_.contains(warns, data.license)) {
      console.error('Warning: '.yellow + 'More informations are required by\
the license. You should check them at the bottom of the file.');
    }

    console.log('License saved to: %s', path.resolve(data.file).green);
  })

  .catch(function (err) {
    console.log('Error: '.red + err.message);
  });


index.js
'use strict';
var _        = require('underscore');
var child    = require('child_process');
var fs       = require('fs');
var inquirer = require('inquirer');
var mkdirp   = require('mkdirp');
var os       = require('os');
var path     = require('path');
var Q        = require('q');
var readline = require('readline');
               require('colors');

// ## Declarations
var api = {};
var licenses = {
  'Apache-2.0': ['name', 'years'],
  'BSD-2-Clause': ['name', 'years'],
  'BSD-3-Clause': ['name', 'years'],
  'CDDL-1.0': [],
  'EPL-1.0': [],
  'GPL-2.0': ['name', 'software', 'description', 'years'],
  'GPL-3.0': ['name', 'software', 'description', 'years'],
  'ISC': ['name', 'email', 'years'],
  'LGPL-2.1': ['name', 'software', 'description', 'years'],
  'LGPL-3.0': [],
  'MIT': ['name', 'email', 'years'],
  'MPL-2.0': [],
};

// ### Intro `lincensr.intro`
// Displays the introduction header.
//
// __Return values__
// * `intro` The introduction text.
var intro = function () {
  return Q.Promise(function (fulfill) {
    var intro = 'Welcome to ' + ('Licensr'.magenta) + '!';
    intro += '\nIf you don\'t know which license is best suited for you visit:'
    intro += '\n * '.grey + 'http://opensource.org/licenses'
    intro += '\n * '.grey + 'http://choosealicense.com/'
    fulfill(intro);
  });
};
api.intro = function (cb) {
  return intro().nodeify(cb);
};


// ### Command
// Run a command in the shell and get the result.
//
// __Arguments__
// * `cmd` The command to run.
//
// __Return values__
// * `stdout` The result of the command.
// * `err` The error as issued by `child_process.exec`.
var command = function (cmd) {
  return Q.Promise(function (fulfill, reject) {
    child.exec(cmd, function (err, stdout) {
      if (err) {
        return reject(err);
      }
      return fulfill(stdout.split('\n')[0]);
    });
  });
};
api.command = function (cmd, cb) {
  return command(cmd).nodeify(cb);
};


// ### Prompt License `lincensr.promptLicense`
// Ask user about the license they want.
//
// __Return values__
// * `licence` A license's name.
var promptLicense = function () {
  return Q.Promise(function (fulfill) {
    var choices = _.keys(licenses);
    inquirer.prompt([{
      type      : 'list',
      name      : 'license',
      message   : 'Which license do you want',
      paginated : true,
      choices   : choices,
      default   : _.indexOf(choices, 'ISC')
    }], function (license) {
      return fulfill(license);
    });
  });
};
api.promptLicense = function (cb) {
  return promptLicense().nodeify(cb);
};

// ### Prompt Info `lincensr.promptInfo`
// Ask user about information required by the license.
//
// __Arguments__
// * `license` A license's name.
// * `infoDefault` Default values issued by a previous command.
//
// __Return values__
// * `answers` Information about the user.
var promptInfo = function (license, infoDefault) {

  if (infoDefault === undefined) {
    infoDefault = {
      name : undefined,
      email: undefined
    };
  }

  return Q.Promise(function (fulfill) {
    var needed    = licenses[license];
    var questions = [];
    // What's your name?
    if (_.contains(needed, 'name')) {
      questions.push({
        type    : 'input',
        name    : 'name',
        message : 'What\'s your name',
        default : infoDefault.name || 'John Doe',
        validate: function (name) {
          return (name.trim().length === 0) ? 'Your name is required' : true;
        }
      });
    }
    // What's your email?
    if (_.contains(needed, 'email')) {
      questions.push({
        type    : 'input',
        name    : 'email',
        message : 'What\'s your email',
        default : infoDefault.email || 'johndoe@acme.org',
        validate: function (email) {
          return (/^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/.test(email) === false)
            ? 'Enter a valid email'
            : true;
        }
      });
    }
    // What's the name of your app?
    if (_.contains(needed, 'software')) {
      questions.push({
        type    : 'input',
        name    : 'software',
        message : 'What\'s the name of your app',
        default : path.basename(path.resolve('.')),
        validate: function (software) {
          return (software.trim().length === 0) ? 'A name is required' : true;
        }
      });
    }
    // Write a brief description of your app
    if (_.contains(needed, 'description')) {
      questions.push({
        type    : 'input',
        name    : 'description',
        message : 'Write a brief description of your app',
        validate: function (description) {
          return (description.trim().length === 0)
            ? 'A description is required'
            : true;
        }
      });
    }
    // On which years does it apply?
    if (_.contains(needed, 'years')) {
      questions.push({
        type    : 'input',
        name    : 'years',
        message : 'On which years does it apply',
        default : new Date().getFullYear(),
        validate: function (years) {
          return (/^\d{4}(-\d{4})?$/.test(years) === false)
            ? 'Enter a four digit year or an range of years'
            : true;
        }
      });
    }
    // Where do you want to save the file?
    questions.push({
      type    : 'input',
      name    : 'file',
      message : 'Where do you want to save the file',
      default : './LICENSE',
    });
    inquirer.prompt(questions, function (answers) {
      return fulfill(answers);
    });
  });
};
api.promptInfo = function (license, infoDefault, cb) {
  return promptInfo(license, infoDefault).nodeify(cb);
};

// ### Write
// Write the license.
//
// __Arguments__
// * `data` The summary fo needed data.
//
// __Return values__
// * `err` If there is any error.
// * `content` Your pimped license!
var write = function (data) {
  return Q.Promise(function (fulfill, reject) {
    var lics = fs.readdirSync(path.resolve(__dirname, 'licenses'));
    if (!_.contains(lics, data.license)) {
      return reject(new Error('Wrong license name'));
    }
    var read = path.resolve(__dirname, 'licenses', data.license);
    var txt = fs.readFileSync(read).toString();
    txt = txt.replace(/%NAME%/g, data.name);
    txt = txt.replace(/%EMAIL%/g, data.email);
    txt = txt.replace(/%SOFTWARE%/g, data.software);
    txt = txt.replace(/%DESCRIPTION%/g, data.description);
    txt = txt.replace(/%YEARS%/g, data.years);
    mkdirp(path.dirname(path.resolve(data.file)), function (err) {
      if (err) {
        return reject(err);
      }
      fs.writeFile(path.resolve(data.file), txt, function (err) {
        if (err) {
          return reject(err);
        }
        return fulfill(txt);
      });
    });
  });
};
api.write = function (data, cb) {
  return write(data).nodeify(cb);
};

// Expose the API
module.exports = api;
